plat/common/include/arm/smccc.h

```diff
@@ -0,0 +1,274 @@
+ * Sets the conduit and version to use when issuing SMCCC calls.
+ *
+ * Platforms must derive the information using platform-specific
+ * methods (device tree / ACPI / other). For version discovery
+ * see also smccc_version().
+ *
+ * If this function is not called, the default values used are:
+ * - conduit: SMC
+ * - version: SMCCC_VERSION_1_2
+ *
+ * @param config Pointer to config struct
+ */
+void smccc_init(struct smccc_config *config);
+
+/**
+ * Issues an SMC.
+ *
+ * Upon completion args are updated with return values
+ * depending on the Function ID and the SMCCC version
+ * implemented.
+ *
+ * @param args Arguments to be passed to the SMC call.
+ *             Updated with return values upon completion.
+ */
+void smccc_smc(struct smccc_args *args);
+
+/**
+ * Issues an HVC
+ *
+ * Upon completion args are updated with return values
+ * depending on the Function ID and the SMCCC version
+ * implemented.
+ *
+ * @param args Arguments to be passed to the HVC call.
+ *             Updated with return values upon completion.
+ */
+void smccc_hvc(struct smccc_args *args);
+
+/**
+ * Issue an SMCCC call using the selected conduit
+ *
+ * Upon completion args are updated with return values
+ * depending on the Function ID and the SMCCC version
+ * implemented.
+ *
+ * @param args Arguments to be passed to the conduit call.
+ *             Updated with return values upon completion.
+ */
+void smccc_invoke(struct smccc_args *args);
+
+/**
+ * Requests the implemented SMCCC version
+ *
+ * This function is optional on SMCCC v1.0 and mandatory
+ * from SMCCC v1.1.
+ *
+ * Notice: Before calling this function one must make sure
+ * that it is safe to do so. For methods to obtain this information
+ * see Appendix B of ARM-DEN0028C.
+ *
+ * @return 32-bit version or SMCCC_NOT_SUPPORTED
+ */
+unsigned long smccc_version(void);
+
+/**
+ * Requests information on the implementation of an
+ * Architecture Service Function
+ *
+ * This function is optional on SMCCC v1.0 and mandatory
+ * from SMCCC v1.1.
+ *
+ * @param fid Function ID of feature to query
+ * @return SMCCC_SUCCESS indicates implemented.
+ *          Other positive values indicated implemented and contain
+ *          additional feature-specific flags.
+ *          A negative value indicates not implemented.
+ */
+unsigned long smccc_arch_features(unsigned long fid);
+
+/**
+ * Obtains the SoC ID, as defined by the silicon provider
+ *
+ * @param type SMCCC_ARCH_SOC_VERSION or SMCCC_ARCH_SOC_REVISION
+ * @return SoC ID on success, NOT_IMPLEMENTED or INVALID_PARAMETER on error
+ */
+unsigned long smccc_arch_soc_id(unsigned long type);
+
+/**
+ * Requests the firmware to apply workaround for CVE-2017-5715
+ *
+ * This function should be called on every PE that requires a firmware
+ * mitigation for CVE-2017-5715. Should be called only when feature
+ * discovery returns a non-negative value. For more information see
+ * ARM DEN0028C Sect. 7.5
+ */
+void smccc_arch_workaround_1(void);
```

Comment:
Originally, I thought that this should be called differently. Perhaps `smccc_arch_workaround_2017_5715`. But then I found that the manual uses this name so it is probably better to keep it although a bit ugly.

Reply:
Yeah, it's how Arm calls it in the spec so I thought it would be a good idea to stick to the standard terminology despite being ugly 👹